generator/default.p9k

```diff
@@ -280,68 +313,202 @@ function __p9k_prompt_custom() {
   done
 
+  # Render End of left prompt
+  __p9k_left_prompt_end
+
+  # Render End of right prompt
   # last whitespace with last segments right symbol, if any
   local last_symbol
   if p9k::defined "P9K_RIGHT_PROMPT_LAST_SEGMENT_END_SYMBOL"; then
     last_symbol="%K{none}%F${CURRENT_RIGHT_BG#%K}${P9K_RIGHT_PROMPT_LAST_SEGMENT_END_SYMBOL}"
   fi
+  __p9k_unsafe[right]+="${__P9K_DATA[LAST_WHITESPACE]}${last_symbol}"
+
+  # Render actual PROMPTS
+  local LC_ALL="" LC_CTYPE="en_US.UTF-8" # Set the right locale to protect special characters
+
+  if [[ "${P9K_PROMPT_ON_NEWLINE:-}" == true ]]; then
+    __p9k_unsafe[left]="${__P9K_ICONS[MULTILINE_FIRST_PROMPT_PREFIX]}%f%b%k${__p9k_unsafe[left]}
+${__P9K_ICONS[MULTILINE_LAST_PROMPT_PREFIX]}"
+    if [[ "$P9K_RPROMPT_ON_NEWLINE" != true ]]; then
+      # The right prompt should be on the same line as the first line of the left
+      # prompt. To do so, there is just a quite ugly workaround: Before zsh draws
+      # the RPROMPT, we advise it, to go one line up. At the end of RPROMPT, we
+      # advise it to go one line down. See:
+      # http://superuser.com/questions/357107/zsh-right-justify-in-ps1
+      RPROMPT_PREFIX='%{'$'\e[1A'${P9K_RPROMPT_PREFIX:-}'%}' # one line up
+      RPROMPT_SUFFIX='%{'${P9K_RPROMPT_SUFFIX:-}$'\e[1B''%}' # one line down
+    else
+      RPROMPT_PREFIX='%{'${P9K_RPROMPT_PREFIX:-}'%}'
+      RPROMPT_SUFFIX='%{'${P9K_RPROMPT_SUFFIX:-}'%}'
+    fi
+  else
+    __p9k_unsafe[left]="%f%b%k${__p9k_unsafe[left]}"
+    RPROMPT_PREFIX='%{'${P9K_RPROMPT_PREFIX:-}'%}'
+    RPROMPT_SUFFIX='%{'${P9K_RPROMPT_SUFFIX:-}'%}'
+  fi
+
+  if [[ "${P9K_DISABLE_RPROMPT:-}" != true ]]; then
+    __p9k_unsafe[right]="${RPROMPT_PREFIX}%f%b%k${__p9k_unsafe[right]}%{${reset_color}%}${RPROMPT_SUFFIX}"
+    RPROMPT='${__p9k_unsafe[right]}'
+  fi
+
+  # Allow iTerm integration to work
+  [[ "${ITERM_SHELL_INTEGRATION_INSTALLED:-}" == "Yes" ]] \
+    && __p9k_unsafe[left]="%{$(iterm2_prompt_mark)%}${__p9k_unsafe[left]}"
+
+local NEWLINE='
+'
+
+  if [[ "${P9K_PROMPT_ADD_NEWLINE:-}" == true ]]; then
+    NEWLINES=""
+    repeat ${P9K_PROMPT_ADD_NEWLINE_COUNT:-1} { NEWLINES+=${NEWLINE} }
+    __p9k_unsafe[left]="$NEWLINES${__p9k_unsafe[left]}"
+  fi
+
+  # By evaluating $__p9k_unsafe[left] here in __p9k_render we avoid
+  # the evaluation of $PROMPT being interrupted.
+  # For security $PROMPT is never set directly. This way the prompt render is
+  # forced to evaluate the variable and the contents of $__p9k_unsafe[left]
+  # are never executed. The same applies to $RPROMPT.
+  PROMPT='${__p9k_unsafe[left]}'
 
-  echo -n "${__P9K_DATA[LAST_WHITESPACE]}${last_symbol}"
+  # About .reset-promt see:
+  # https://github.com/sorin-ionescu/prezto/issues/1026
+  # https://github.com/zsh-users/zsh-autosuggestions/issues/107#issuecomment-183824034
+  [[ "${1}" == "true" ]] && zle .reset-prompt
```

Comment:
`zle -R` (force redisplay) should be added at the end. Otherwise, the current prompt may be moved one line up and overwrite previous prompt.

```zsh
  [[ "${1}" == "true" ]] && zle .reset-prompt && zle -R
```